### DISCUSSION TITLE
Extending next/dynamic to only load client bundle when in viewport

### DISCUSSION THREAD

**Discussion by alecrae** [vercel/next.js#27941] [Ideas]
I have been utilising the `next/dynamic` import pattern with success in performance scores, and had an idea on how to improve it further to reduce initial JS payload size. While it currently improves the behaviour of JS loading, the JS is still loaded asap when it may not be needed immediately.  So there could be an optional argument for `dynamic` to only load the client-side JS for that component when the SSR'd component is close to being inside the viewport, similar to how lazily loaded images are loaded when they are close being in the viewport.

An example use case would be a newsletter signup form in a footer, with the form component pulling in libraries like formik and yup for validation. It could be rendered SSR, so CLS score, a11y etc is maintained, and once the user scrolls near the bottom of the page, the JS is loaded and component is hydrated. 

```
const NewsletterForm = dynamic(() => import("src/newsletter-form"), {
  loadWhenInView: true
});
```

I'm not sure if this approach would fit with React's hydration modal, or if there is an existing solution for this, but thought i would put this out there for feedback.

**Comment by deadcoder0904** [comment]
Great idea. Benefits over `IntersectionObserver` to achieve the same thing? I specifically use https://github.com/thebuilder/react-intersection-observer & it [works pretty well](https://github.com/thebuilder/react-intersection-observer/discussions/486) :)

**Reply by alecrae** [reply]
I was thinking the implementation would use something like `IntersectionObserver` to achieve it. I agree that library is great and quite small too

**Reply by deadcoder0904** [reply]
Yeah, built-in way would be great with additional offset parameters (x, y) as well.
